homeassistant/components/nibe_heatpump/__init__.py

```diff
@@ -0,0 +1,242 @@
+from nibe.heatpump import HeatPump, Model
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_IP_ADDRESS, CONF_MODEL, Platform
+from homeassistant.core import CALLBACK_TYPE, HomeAssistant
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers import device_registry as dr
+from homeassistant.helpers.entity import DeviceInfo, async_generate_entity_id
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
+)
+
+from .const import (
+    CONF_CONNECTION_TYPE,
+    CONF_CONNECTION_TYPE_NIBEGW,
+    CONF_LISTENING_PORT,
+    CONF_REMOTE_READ_PORT,
+    CONF_REMOTE_WRITE_PORT,
+    CONF_WORD_SWAP,
+    DOMAIN,
+)
+from .utils import TooManyTriesException, retry
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORMS: list[Platform] = [Platform.SENSOR]
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Nibe Heat Pump from a config entry."""
+
+    heatpump = HeatPump(Model[entry.data[CONF_MODEL]])
+    heatpump.word_swap = entry.data[CONF_WORD_SWAP]
+    heatpump.initialize()
+
+    connection_type = entry.data[CONF_CONNECTION_TYPE]
+
+    if connection_type == CONF_CONNECTION_TYPE_NIBEGW:
+        connection = NibeGW(
+            heatpump,
+            entry.data[CONF_IP_ADDRESS],
+            entry.data[CONF_REMOTE_READ_PORT],
+            entry.data[CONF_REMOTE_WRITE_PORT],
+            listening_port=entry.data[CONF_LISTENING_PORT],
+        )
+    else:
+        raise HomeAssistantError(f"Connection type {connection_type} is not supported.")
+
+    await connection.start()
+    try:
+        coordinator = Coordinator(hass, heatpump, connection)
+
+        data = hass.data.setdefault(DOMAIN, {})
+        data[entry.entry_id] = coordinator
+        await coordinator.async_config_entry_first_refresh()
+
+        reg = dr.async_get(hass)
+        reg.async_get_or_create(
+            config_entry_id=entry.entry_id,
+            identifiers={(DOMAIN, entry.unique_id or entry.entry_id)},
+            manufacturer="NIBE Energy Systems",
+            model=heatpump.model.name,
+            name=heatpump.model.name,
+        )
+
+        await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+
+    except Exception:
+        await connection.stop()
+        raise
+
+    # Trigger a refresh again now that all platforms have registered
+    hass.async_add_job(coordinator.async_refresh)
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
+        coordinator: Coordinator = hass.data[DOMAIN].pop(entry.entry_id)
+        await coordinator.connection.stop()
+
+        # sleep a bit to make sure sockets are closed
+        await asyncio.sleep(5)
```

Comment:
Is there not a more explicitly way to detect this? Additionally, what if sockets are still open after 5 seconds for some reason?

Reply:
I don't know any way. I think it might be enough to yield, but unsure. asyncio core will add a call_soon task to close the sockets fully. I don't know any way to keep track of how long that takes. I will look thou.

Reply:
sleep(0) is likely enough. I've opened a pull on upstream lib to include that in the stop command.

Reply:
And I merged that change to the lib.